### ISSUE TITLE
Link to attached file

### ISSUE THREAD

**Issue by aron89682** [pikepdf/pikepdf#279]
Is it possible to add a link that would be able to open an attached file?

**Comment by jbarlow83** [comment]
Yes it's possible. The PDF RM section 12.6.4.4 describes the Action structure you need to do this and pikepdf can create them all. Unfortunately it does not currently provide a high level wrapper for this task. You'd then need to create something to trigger that action, such as by creating an annotation (e.g. a clickable button) in section 12.5.

If you're generating a PDF from scratch you might find a higher level PDF library easier to work with. There are some "write only" PDF libraries like WeasyPrint and reportlab that have more output generating features. pikepdf shines when you need to edit, process or fix existing PDFs.

**Comment by aron89682** [comment]
Thanks, I'll check out one of the other libraries.

**Comment by aron89682** [comment]
@jbarlow83 Are there any plans on adding a wrapper for Actions?  I checked the other libraries and so far this is the best one for adding the images to an already existing PDF, that I've found.  I'm just struggling with creating the Action.

I will receive pre-generated PDF's and will need to add multiple attachments to it and then link to those attachments.  You mentioned that I would need to create and Action and then an Annotation pointing to the Action.  I'm a little confused where/how to add the Action structure, could you help me understand or point me in the right direction?

I'm currently doing something like this to add the attachments:
```
filespec = AttachedFileSpec(new_pdf, data=base64.b64decode(bytes), filename='image01.jpg')
new_pdf.attachments['image01.jpg'] = filespec
```

Do I need to add it to the filespec or do I need to create a completely separate Dictionary Object?

Below is from the PDF RM section 12.6.4.4 Embedded Go-To Actions:
```
1 0 obj
     << /Type /Action
          /S /GoToE
          /D (Chapter 1)
          /T     << /R /C
                        /N (Embedded document) >>
     >>
endobj
```

**Comment by jbarlow83** [comment]
Unfortunately - probably not. There's a lot of actions.

I also don't know if Acrobat actually supports the concept of a _link_ that opens a file that is attached to the PDF. At least I can't figure out how to do it through the GUI, which doesn't bode well. It looks like it may be an option only when the attached file is another PDF.

You can insert an annotation object with a customizable appearance (default is push pin) and arrange for clicking that to open an embedded file. I update the documentation to show how to do this.

Incidentally, if you have access to Acrobat, creating the feature you want in a trivial PDF and then checking the description of the object in Preflight > Browse Internal Document Structure is often the easiest way to find out how to create what you want.

**Comment by aron89682** [comment]
Thanks for the Preflight thought.  That seems to be showing a FileAttachment Annotation (12.5.6.15).

![image](https://user-images.githubusercontent.com/76072805/146095161-94dd4bd3-e063-4af2-b518-0b6cbed7a9dc.png)
